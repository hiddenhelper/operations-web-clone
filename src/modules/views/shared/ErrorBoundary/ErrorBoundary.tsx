import React from 'react';

export interface IErrorBoundaryProps {
  Logger: { error: (...args) => void };
  fallback?: React.ReactElement;
}

export interface IErrorBoundaryState {
  hasError: boolean;
}

export default class ErrorBoundary extends React.PureComponent<IErrorBoundaryProps, IErrorBoundaryState> {
  public static defaultProps = {
    Logger: null,
    children: null,
    fallback: null,
  };

  public state = { hasError: false };

  public static getDerivedStateFromError = (error: Error): IErrorBoundaryState => ({ hasError: true });

  public componentDidCatch(error, info) {
    const { Logger } = this.props;
    /* istanbul ignore else */
    if (Logger.error) Logger.error('ErrorBoundary: ', error, info);
  }

  public render() {
    const { children, fallback } = this.props;
    const { hasError } = this.state;

    return hasError ? fallback : children;
  }
}
