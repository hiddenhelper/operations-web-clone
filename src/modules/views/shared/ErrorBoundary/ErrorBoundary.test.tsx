import React from 'react';
import { create } from 'react-test-renderer';

import ErrorBoundary from './ErrorBoundary';

class HandGranade extends React.Component {
  public render() {
    const a: any = null;
    a.b.c = 10;
    return <div>This should not work</div>;
  }
}

const Logger = { error: jest.fn() };

describe('ErrorBoundaryTest', () => {
  global.console.error = () => null;
  it('renders child', () => {
    expect(
      create(
        <ErrorBoundary>
          <div>Render test</div>
        </ErrorBoundary>
      ).toJSON()
    ).toMatchSnapshot();
  });

  it('renders placeholder when child render fails', () => {
    expect(
      create(
        <ErrorBoundary Logger={Logger} fallback={<h1>Error!</h1>}>
          <HandGranade>Render test</HandGranade>
        </ErrorBoundary>
      ).toJSON()
    ).toMatchSnapshot();
    expect(Logger.error).toHaveBeenCalled();
  });
});
