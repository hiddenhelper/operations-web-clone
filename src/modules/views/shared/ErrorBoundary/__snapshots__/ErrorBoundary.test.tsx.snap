// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundaryTest renders child 1`] = `
<div>
  Render test
</div>
`;

exports[`ErrorBoundaryTest renders placeholder when child render fails 1`] = `
<h1>
  Error!
</h1>
`;
